### PR TITLE
[Feat]Add the features for expanding and shrinking the number of tables in distributed training by independently saving files.

### PR DESCRIPTION
**Pull request by MoFHeka**
# Description

Add the features for expanding and shrinking the number of tables in distributed training by independently saving files.

Also improve the performance of CPU table by using std::copy_n.

Also make genarating _DEFAULT_CUDA_COMPUTE_CAPABILITIES more compatible and concise in build_deps/toolchains/gpu/cuda_configure.bzl.

Also compatible with TF 2.9, which would pass parameter validate_shape to _init_from_args.

Also fix RedisTableOfTensors Node missing user-defined name.

Also fix problem with the parameter 'checkpoint' passing not working when using DE BasicEmbedding.

Also compatible with 'find_namespace_packages' when using setuptools, because 'find_packages' has been deprecated.

## Type of change

- [ ] Bug fix
- [x] New Tutorial
- [x] Updated or additional documentation
- [x] Additional Testing
- [x] New Feature

# Checklist:

- [x] I've properly [formatted my code according to the guidelines](https://github.com/tensorflow/recommenders-addons/blob/master/CONTRIBUTING.md#coding-style)
    - [x] By running yapf
    - [x] By running clang-format
- [ ] This PR addresses an already submitted issue for TensorFlow Recommenders-Addons
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works

# How Has This Been Tested?
Read the doc(docs/api_docs/tfra/dynamic_embedding/FileSystemSaver.md) and run new tests.

### CODE REVIEW

File: tensorflow_recommenders_addons/dynamic_embedding/python/kernel_tests/dynamic_embedding_variable_test.py [hunk 1]
```diff
@@ -59,6 +59,8 @@
 from tensorflow.python.training.tracking import util as track_util
 from tensorflow.python.util import compat
 
+import tensorflow as tf
```
[the remark this answers is not in the document]
Comment (MoFHeka) [reply tensorflow_recommenders_addons/dynamic_embedding/python/kernel_tests/dynamic_embedding_variable_test.py]: > Please check this change before: https://github.com/tensorflow/recommenders-addons/pull/284/files#diff-4077cebe6be4ca369d86f091a6e337e530e69cfc2b58dacdc27cb18b813c2cdaR33-L42, `import tf` may not the ideal way to resolve the issue?

fix

File: tensorflow_recommenders_addons/dynamic_embedding/python/ops/dynamic_embedding_creator.py [hunk 2]
```diff
@@ -225,3 +246,129 @@ def get_config(self):
...
+
+# SaveableObjectCreator
+class SaveableObjectCreator(object, metaclass=ABCMeta):
+  """
+      
+
+    Example usage:
+
+    ```python
+    saveable_config=
```
Comment (rhdong) [inline tensorflow_recommenders_addons/dynamic_embedding/python/ops/dynamic_embedding_creator.py:259]: It seems not to be finished.

File: tensorflow_recommenders_addons/dynamic_embedding/python/ops/dynamic_embedding_variable.py [hunk 3]
```diff
@@ -131,6 +138,194 @@ def default_partition_fn(keys, shard_num):
...
+    def insert_de_table(insert_keys_tensor_array_list,
+                        insert_values_tensor_array_list):
+      with ops.name_scope(de_variable.name, "%s_table_restore" % shard._name):
+        all_keys_partitions_tensor_array = insert_keys_tensor_array_list[
+            shard_idx]
+        all_values_partitions_tensor_array = insert_values_tensor_array_list[
+            shard_idx]
+
+        def sub_saveable_insert(i):
+          insert_op = shard.insert(all_keys_partitions_tensor_array.read(i),
```
Comment (rhdong) [inline tensorflow_recommenders_addons/dynamic_embedding/python/ops/dynamic_embedding_variable.py:306]: If this insert operation will cause the OOM on some workers? Do we have evaluated the risk here?

Comment (MoFHeka) [reply tensorflow_recommenders_addons/dynamic_embedding/python/ops/dynamic_embedding_variable.py:306]: > If this insert operation will cause the OOM on some workers? Do we have evaluated the risk here?

Yes, of course a big risk, and I don't have a good solution.
Besides this, there is no read_byte OP in TF, so I have to read the whole KV file which also may cause being OOM Killed.

Comment (MoFHeka) [reply tensorflow_recommenders_addons/dynamic_embedding/python/ops/dynamic_embedding_variable.py]: Already fix this by using TF FixedLengthRecordDataset to read buffer.

File: tensorflow_recommenders_addons/dynamic_embedding/python/ops/dynamic_embedding_creator.py (hunk 2 above)
Comment (rhdong) [reply tensorflow_recommenders_addons/dynamic_embedding/python/ops/dynamic_embedding_creator.py:256]: @MoFHeka

File: tensorflow_recommenders_addons/dynamic_embedding/python/ops/dynamic_embedding_variable.py [hunk 4]
```diff
@@ -621,6 +906,115 @@ def export(self, name=None):
         full_values.append(vals_)
     return array_ops.concat(full_keys, 0), array_ops.concat(full_values, 0)
 
+  def save_to_file_system(self,
+                          dirpath,
+                          mpi_size=1,
+                          mpi_rank=0,
+                          file_name_list=[],
```
Comment (rhdong) [inline tensorflow_recommenders_addons/dynamic_embedding/python/ops/dynamic_embedding_variable.py:913]: Just use None for empty, or all of the callings will share the `[]`.

**Comment by rhdong** [comment]
Highly recommend squashing all of the commits for compatibility with TF2.9, and squashing the last two commits as I guess there's one commit that can't pass the CI tests, I mean this one(we'd better guarantee each commit can pass CI separately): [[fix] Fixed problem with the parameter 'checkpoint' passing not worki](https://github.com/tensorflow/recommenders-addons/pull/290/commits/2afd1f31e5b6869e6cc2fa4ea422e80c3221bed9).

File: docs/api_docs/tfra/dynamic_embedding/FileSystemSaveableObjectCreator.md [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+import tensorflow_recommenders_addons.dynamic_embedding as de
+
+test_devices = ['GPU:0', 'GPU:1']
+params0 = de.get_variable(
+            'table0',
+            devices=test_devices,
+            dim=8,
+            initializer=0.0,
+            bp_v2=True,
+            saveable_object_creator=de.FileSystemSaveableObjectCreator())
```
Comment (rhdong) [inline docs/api_docs/tfra/dynamic_embedding/FileSystemSaveableObjectCreator.md:66]: There are not too many comments for me here, I just want to know if we have better choice for the parameter, this requires the user to understand the internal mechanism of `SaveableObject`, it's too professional maybe. If we can change it to `saver` and `saverCreater` and `ShardSaver`?

File: tensorflow_recommenders_addons/dynamic_embedding/python/ops/dynamic_embedding_creator.py [hunk 6]
```diff
@@ -225,3 +243,150 @@ def get_config(self):
...
+                                   shard_idx,
+                                   name,
+                                   full_name=""):
+    raise NotImplementedError('create function must be implemented')
+
+  def set_upsert_restore(self, setting):
+    self._upsert_restore = setting
+
+
+class FileSystemSaveableObjectConfig(object):
```
Comment (rhdong) [inline tensorflow_recommenders_addons/dynamic_embedding/python/ops/dynamic_embedding_creator.py:279]: Considering the Config class only serves for MPI/HVD, we may name it `DistributedFileSystem`？ If your initial idea is not only for MPI scenarios, may we need to change it to a common name that will benefit extending in the future?

File: docs/api_docs/tfra/dynamic_embedding/FileSystemSaveableObjectCreator.md (hunk 5 above)
Comment (MoFHeka) [reply docs/api_docs/tfra/dynamic_embedding/FileSystemSaveableObjectCreator.md:66]: saver_creator?

File: tensorflow_recommenders_addons/dynamic_embedding/python/ops/dynamic_embedding_variable.py (hunk 4 above)
Comment (MoFHeka) [reply tensorflow_recommenders_addons/dynamic_embedding/python/ops/dynamic_embedding_variable.py:913]: Done

File: tensorflow_recommenders_addons/dynamic_embedding/core/kernels/lookup_impl/lookup_table_op_cpu.h [hunk 7]
```diff
@@ -287,36 +282,41 @@ class TableWrapperDefault final : public TableWrapperBase<K, V> {
...
     ValueType value_vec;
+    value_vec.reserve(value_dim);
     for (int64 j = 0; j < value_dim; j++) {
       V value = value_flat(index, j);
       value_vec.push_back(value);
     }
     return table_->insert_or_assign(key, value_vec);
   }
 
+  bool insert_or_assign_one(K* key, V* value, int64 value_dim) const override {
```
Comment (rhdong) [inline tensorflow_recommenders_addons/dynamic_embedding/core/kernels/lookup_impl/lookup_table_op_cpu.h:293]: `insert_or_assign_one` -> `insert_or_assign`

File: tensorflow_recommenders_addons/dynamic_embedding/python/ops/dynamic_embedding_creator.py [hunk 8]
```diff
@@ -225,3 +243,150 @@ def get_config(self):
...
+
+  def __init__(self,
+               mpi_size: int = None,
```
Comment (rhdong) [inline tensorflow_recommenders_addons/dynamic_embedding/python/ops/dynamic_embedding_creator.py:282]: `table_id` or `table_size`

File: tensorflow_recommenders_addons/dynamic_embedding/python/ops/dynamic_embedding_creator.py (hunk 6 above)
Comment (MoFHeka) [reply tensorflow_recommenders_addons/dynamic_embedding/python/ops/dynamic_embedding_creator.py]: This FileSystem class not only for MPI/HVD, it is also able to be used by TF origin training code. It make user easy to use filesystem for storing DE KV without any additional code by patching TF.

File: tensorflow_recommenders_addons/dynamic_embedding/core/kernels/lookup_impl/lookup_table_op_cpu.h (hunk 7 above)
Comment (MoFHeka) [reply tensorflow_recommenders_addons/dynamic_embedding/core/kernels/lookup_impl/lookup_table_op_cpu.h]: Done.

File: tensorflow_recommenders_addons/dynamic_embedding/python/ops/dynamic_embedding_creator.py (hunk 8 above)
Comment (MoFHeka) [reply tensorflow_recommenders_addons/dynamic_embedding/python/ops/dynamic_embedding_creator.py]: proc_rank and proc_size following TF style?
